### PR TITLE
Make sure subscription completed before testing

### PR DESCRIPTION
**Pull request by johanandren**
References #28885

Fixes tests that were subscribing to event bus for Dropped/DeadLetter etc but didn't verify that the subscription had completed before expecting to be able to see published messages.

There were a few more tests using the event bus that either did verify using debug logs or had lots of other operations in between that I did not change.

### CODE REVIEW

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorContextSpec.scala [hunk 1]
```diff
@@ -88,6 +89,13 @@ abstract class ActorContextSpec extends ScalaTestWithActorTestKit with AnyWordSp
     "canonicalize behaviors" in {
       val unhandledProbe = createTestProbe[UnhandledMessage]()
       system.eventStream ! EventStream.Subscribe(unhandledProbe.ref)
+      // make sure subscription completed before moving on
+      unhandledProbe.awaitAssert {
+        val uh = UnhandledMessage("dummy", akka.actor.ActorRef.noSender, unhandledProbe.ref.toClassic)
+        system.eventStream ! EventStream.Publish(uh)
+        unhandledProbe.expectMessageType[UnhandledMessage] shouldBe theSameInstanceAs(uh)
+      }
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorContextSpec.scala:97]: Searching for "eventStream ! EventStream.Subscribe" shows that we have such testing in many places. Would it be worth adding some testkit utility for this?
```
val unhandledProbe: TestProbe[UnhandledMessage] = testKit.createUnhandledMessageProbe()
```

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala [hunk 2]
```diff
@@ -681,6 +689,13 @@ class UnstashingSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike with
     "deal with stop" in {
       val probe = TestProbe[Any]
       system.eventStream ! EventStream.Subscribe(probe.ref.narrow[DeadLetter])
+      // make sure subscription completed before moving on
+      probe.awaitAssert {
+        import akka.actor.typed.scaladsl.adapter._
+        val dl = DeadLetter("dummy", probe.ref.toClassic, probe.ref.toClassic)
+        system.eventStream ! EventStream.Publish(dl)
+        probe.expectMessageType[DeadLetter] shouldBe theSameInstanceAs(dl)
+      }
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala:698]: and similar testkit util for Dropped and DeadLetter

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala [hunk 3]
```diff
@@ -681,6 +689,13 @@ class UnstashingSpec extends ScalaTestWithActorTestKit with AnyWordSpecLike with
     "deal with stop" in {
       val probe = TestProbe[Any]
       system.eventStream ! EventStream.Subscribe(probe.ref.narrow[DeadLetter])
+      // make sure subscription completed before moving on
+      probe.awaitAssert {
+        import akka.actor.typed.scaladsl.adapter._
+        val dl = DeadLetter("dummy", probe.ref.toClassic, probe.ref.toClassic)
```
Comment (patriknw) [inline akka-actor-typed-tests/src/test/scala/akka/actor/typed/scaladsl/StashSpec.scala:695]: those dummy messages should implement DeadLetterSuppression to avoid logging

File: akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorContextSpec.scala (hunk 1 above)
Comment (johanandren) [reply akka-actor-typed-tests/src/test/scala/akka/actor/typed/ActorContextSpec.scala:97]: Mmm, not a bad idea. I thought a bit about extending the event stream protocol with an subscribe+ack while working on this one. But thinking about it again it pretty much only matters in a test context where you have a single thread subscribing and then publishing/asserting.

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ActorTestKit.scala [hunk 4]
```diff
@@ -217,6 +220,25 @@ final class ActorTestKit private[akka] (delegate: akka.actor.testkit.typed.scala
...
+   *         will be completed and verified so any dropped letter after it will be caught by the probe.
+   */
+  def createDroppedMessageProbe(): TestProbe[Dropped] =
+    delegate.createDroppedMessageProbe().asJava
+
+  /**
+   * @return A test probe that is subscribed to dead letters from the system event bus. Subscription
+   *         will be completed and verified so any dead letter after it will be caught by the probe.
+   */
+  def createDeadLetterMessageProbe(): TestProbe[DeadLetter] = delegate.createDeadLetterMessageProbe().asJava
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/javadsl/ActorTestKit.scala:234]: maybe just createDeadLetterProbe

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala [hunk 5]
```diff
@@ -215,6 +223,63 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
...
+  def createUnhandledMessageProbe(): TestProbe[UnhandledMessage] = {
+    val probe = createTestProbe[Any]()
+    val narrowed = probe.ref.narrow[UnhandledMessage]
+    system.eventStream ! EventStream.Subscribe(narrowed)
+
+    // publish-receive roundtrip to verify we are subscribed before returning the probe
+    probe.awaitAssert {
+      val uh = UnhandledMessage(ActorTestKit.dummyMessage, probe.ref.toClassic, probe.ref.toClassic)
+      system.eventStream ! EventStream.Publish(uh)
+      if (probe.expectMessageType[UnhandledMessage].message != ActorTestKit.dummyMessage)
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:239]: should it fish in case there are some other concurrent UnhandledMessage?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala [hunk 6]
```diff
@@ -215,6 +223,63 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
    */
   def createTestProbe[M](name: String): TestProbe[M] = TestProbe(name)(system)
 
+  /**
+   * @return A test probe that is subscribed to unhandled messages from the system event bus. Subscription
+   *         will be completed and verified so any unhandled message after it will be caught by the probe.
+   */
+  def createUnhandledMessageProbe(): TestProbe[UnhandledMessage] = {
+    val probe = createTestProbe[Any]()
+    val narrowed = probe.ref.narrow[UnhandledMessage]
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:232]: why is this needed, instead of `createTestProbe[UnhandledMessage]()`?

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala [hunk 7]
```diff
@@ -215,6 +223,63 @@ final class ActorTestKit private[akka] (val name: String, val config: Config, se
...
+
+    // publish-receive roundtrip to verify we are subscribed before returning the probe
+    probe.awaitAssert {
+      val dl = Dropped(ActorTestKit.dummyMessage, "no reason", probe.ref.toClassic, probe.ref.toClassic)
+      system.eventStream ! EventStream.Publish(dl)
+      if (probe.expectMessageType[Dropped].message != ActorTestKit.dummyMessage)
+        throw new RuntimeException("Not the expected unhandled")
+    }
+    probe.asInstanceOf[TestProbe[Dropped]]
+  }
```
Comment (patriknw) [inline akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:281]: Everything the same for these 3 apart for the published message. Should be possible to DRY.

**Review by patriknw — COMMENTED** [review]
good addition to testkit

File: akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala (hunk 6 above)
Comment (johanandren) [reply akka-actor-testkit-typed/src/main/scala/akka/actor/testkit/typed/scaladsl/ActorTestKit.scala:232]: Half done thought about having some completely different message to send across bus for verification that subscription completed but I scrapped that and forgot to change.

**Review by patriknw — APPROVED** [review]
LGTM

**Review by octonato — APPROVED** [review]
